README
Note about Gtk3 : 

In order to load interface glade interface, gtk.glade cannot be used anymore, it must be replace with a Gtk.Builder.
For this to work, the .glade file needs to be converted to a .ui file.
This can be done using :

   gtk-builder-convert filename.glade filename.ui


debian/changelog
mintnanny (1.6.0) vanessa; urgency=medium

  * Bump rebuild without ZST

 -- Clement Lefebvre <root@linuxmint.com>  Thu, 04 Aug 2022 11:54:26 +0200

mintnanny (1.5.9) vanessa; urgency=medium

  [ Redecoffee ]
  * Reject domains which length is larger than 63 characters (#18)

  [ Clement Lefebvre ]
  * l10n: Update files

 -- Clement Lefebvre <root@linuxmint.com>  Fri, 10 Jun 2022 11:27:54 +0200

mintnanny (1.5.8) ulyana; urgency=medium

  * l10n: Generate files

 -- Clement Lefebvre <root@linuxmint.com>  Thu, 14 May 2020 08:19:39 +0100

mintnanny (1.5.7) tina; urgency=medium

  * Fix generate_desktop_files

 -- Clement Lefebvre <root@linuxmint.com>  Sun, 30 Jun 2019 15:52:27 +0200

mintnanny (1.5.6) tara; urgency=medium

  * Add msg to polkit policy

 -- Clement Lefebvre <root@linuxmint.com>  Wed, 30 May 2018 18:33:02 +0100

mintnanny (1.5.5) tara; urgency=medium

  [ dg1727 ]
  * Added pkexec support

  [ Clement Lefebvre ]
  * Fix makepot, remove msgids from polkit policy, move pkexec to launchers

 -- Clement Lefebvre <root@linuxmint.com>  Mon, 07 May 2018 12:30:10 +0100

mintnanny (1.5.4) sonya; urgency=medium

  * l10n: Generate desktop files

 -- Clement Lefebvre <root@linuxmint.com>  Sun, 07 May 2017 13:06:46 +0100

mintnanny (1.5.3) sonya; urgency=medium

  * Provide a hicolor icon

 -- Clement Lefebvre <root@linuxmint.com>  Mon, 01 May 2017 19:35:07 +0100

mintnanny (1.5.2) sarah; urgency=medium

  [ Daniel Alley ]
  * Replaced deprecated Vbox widget
  * Added gi.require_version('3.0')
  * Removed try/except

  [ Clement Lefebvre ]
  * Updated generated files

 -- Clement Lefebvre <root@linuxmint.com>  Fri, 22 Apr 2016 10:18:46 +0100

mintnanny (1.5.1) sarah; urgency=medium

  [ Daniel Alley ]
  * removed last remnants of Python2
  * fix out of date debian/control file
  * formating fixes

 -- Clement Lefebvre <root@linuxmint.com>  Fri, 19 Feb 2016 12:08:41 +0000

mintnanny (1.5.0) sarah; urgency=medium

  * Fixed www prefix and fugly dialog with GTK 3.18

 -- Clement Lefebvre <root@linuxmint.com>  Wed, 20 Jan 2016 19:21:39 +0000

mintnanny (1.4.9) sarah; urgency=medium

  * Get rid of the commands module
  * Indentation: Convert tabs to spaces
  * Migrated to GTK3
  * Moved icon to /usr/share/pixmaps, moved ui file to /usr/share, decapitilized directories/files
  * Replace mint-common's entrydialog with a GTK3 dialog
  * Makepot
  * Python3, use of class, multiple small fixes
  * Adding domains: validate when pressing Enter
  * Automatically add the www subdomain.
  * Only add the www subdomain for domains with no subdomains

 -- Clement Lefebvre <root@linuxmint.com>  Wed, 20 Jan 2016 19:00:44 +0000

mintnanny (1.4.8) rosa; urgency=medium

  [ Christophe Zeitouny ]
  * Standardize indentation to tabs as that seemed to be what most lines are using. Consistency is what's most important here.
  * Perform sanity checks on the given domain before trying to put it in /etc/hosts. I was able to ruin my hosts file by passing in quotes, double quotes, etc... Plus since the string is pasted as-is into a command that's executed with root privileges, this may have resulted in arbitrary execution of commands. The new code makes sure that the given domain name conforms to MOST restrictions on such names, and displays a helpful error message in case the given name is not well-formed

  [ Clement Lefebvre ]
  * l10n and simplify strings introduced in last PR
  * Updated pot file
  * Updated desktop files

 -- Clement Lefebvre <root@linuxmint.com>  Fri, 06 Nov 2015 17:31:09 +0000

mintnanny (1.4.7) rebecca; urgency=medium

  * Updated debian files

 -- Clement Lefebvre <root@linuxmint.com>  Fri, 17 Oct 2014 11:08:48 +0200

mintnanny (1.4.6) qiana; urgency=low

  * [44225ab] Add trailing semicolon
  * [22bbc84] Update translations

 -- Frédéric Gaudet <fred@linuxmint.com>  Mon, 26 May 2014 13:39:46 +0200

mintnanny (1.4.5) qiana; urgency=low

  * [91a3a41] KDE desktop file generation
  * [16335c3] GNOME based desktop file hidden in KDE
  * [a2533f8] New KDE4 desktop file

 -- Frédéric Gaudet <fred@linuxmint.com>  Mon, 26 May 2014 11:42:07 +0200

mintnanny (1.4.4) qiana; urgency=medium

  * Updated desktop file

 -- Clement Lefebvre <root@linuxmint.com>  Sun, 04 May 2014 14:25:30 +0100

mintnanny (1.4.3) petra; urgency=low

  * Fix desktop file

 -- Frédéric Gaudet <fred@linuxmint.com>  Fri, 08 Nov 2013 09:50:01 +0100

mintnanny (1.4.2) petra; urgency=low

  * Changed icon
  * Switch from 0.0.0.0 to 127.0.0.1

 -- Clement Lefebvre <root@linuxmint.com>  Sat, 19 Oct 2013 20:39:03 +0100

mintnanny (1.4.1) olivia; urgency=low

  * Revert back to GTK2

 -- Clement Lefebvre <root@linuxmint.com>  Tue, 12 Mar 2013 11:31:29 +0000

mintnanny (1.3.9) maya; urgency=low

  * Replaced zenity dependency with mint-common's entrydialog

 -- Clement Lefebvre <root@linuxmint.com>  Mon, 30 Jul 2012 14:34:55 +0100

mintnanny (1.3.8) helena; urgency=low

  * Bug fix: redirecting warning/errors to /dev/null when using zenity

 -- Clement Lefebvre <root@linuxmint.com>  Sat,  14 Aug 2010 14:49:00 +0000

mintnanny (1.3.7) helena; urgency=low

  * Updated desktop file

 -- Clement Lefebvre <root@linuxmint.com>  Fri,  14 May 2010 10:30:00 +0000

mintnanny (1.3.6) helena; urgency=low

  * Changed translation framework

 -- Clement Lefebvre <root@linuxmint.com>  Tue,  3 Nov 2009 11:42:00 +0000

mintnanny (1.3.5) helena; urgency=low

  * Dependency on mint-common

 -- Clement Lefebvre <root@linuxmint.com>  Thu,  27 Jul 2009 17:38:00 +0000

mintnanny (1.3.4) helena; urgency=low

  * Blueprint implemented: https://blueprints.launchpad.net/linuxmint/+spec/generic-branding

 -- Clement Lefebvre <root@linuxmint.com>  Thu,  27 Jul 2009 17:38:00 +0000

mintnanny (1.3.3) gloria; urgency=low

  * Bug fix: No label under toolbar icons (https://bugs.launchpad.net/bugs/399705)

 -- Clement Lefebvre <root@linuxmint.com>  Thu,  16 Jul 2009 12:34:00 +0000

mintnanny (1.3.2) gloria; urgency=low

  * Updated translations


usr/lib/linuxmint/mintNanny/mintNanny.py
#!/usr/bin/python2

# MintNanny
#   Clement Lefebvre <clem@linuxmint.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; Version 2
# of the License.

import os
import sys
from gi.repository import Gtk, Gdk
import gettext
import re
import subprocess

# i18n
gettext.install("mintnanny", "/usr/share/linuxmint/locale")

def add_domain(widget, treeview_domains, window):
    output, error = subprocess.Popen(['/usr/lib/linuxmint/common/entrydialog.py', _("Please type the domain name you want to block"), _("Domain name:"), '', 'mintNanny'], stdout=subprocess.PIPE).communicate()
    domain = re.sub(r'\s', '', output)

    if domain == '':
        # Take no action on empty input
        return

    if not is_valid_domain(domain):
        # User has passed an invalid domain (one that contains invalid characters)
        # Display an error dialog to inform them why we're not adding it to the list
        dlg = Gtk.MessageDialog(type=Gtk.MessageType.ERROR, buttons=Gtk.ButtonsType.OK, message_format=_("Invalid Domain"))
        dlg.set_transient_for(window)
        desc1 = _("'%s' is not a valid domain name." % domain)
        desc2 = _("Domain names must start and end with a letter or a digit, and can only contain letters, digits, dots and hyphens.")
        desc3 = _("Example: my.number1domain.com")
        dlg.format_secondary_text("%s\n\n%s\n\n%s" % (desc1, desc2, desc3))
        dlg.run()
        dlg.destroy()
        return

    model = treeview_domains.get_model()
    iter = model.insert_before(None, None)
    model.set_value(iter, 0, domain)
    domain = "127.0.0.1 " + domain + "  # blocked by mintNanny"
    os.system("echo \"" + domain + "\" >> /etc/hosts")

def remove_domain(widget, treeview_domains):
    selection = treeview_domains.get_selection()
    (model, iter) = selection.get_selected()
    if (iter != None):
        domain = model.get_value(iter, 0)
        os.system("sed '/" + domain + "/ d' /etc/hosts > /tmp/hosts.mintNanny")
        os.system("mv /tmp/hosts.mintNanny /etc/hosts")
        model.remove(iter)

def is_valid_domain(domain):
    #Quick sanity check
    if domain == '':
        return False

    # The following is based on RFC 952 (https://tools.ietf.org/html/rfc952)
    # and section 2.1 of RFC 1123 (https://tools.ietf.org/html/rfc1123#page-13)
    # Also see sections 2.3.1 and 2.3.4 of RFC 1035 (http://tools.ietf.org/html/rfc1035)

    # Quick regex match to check the domain name's sanity
    # Note: This enforces that the domain name starts with a letter or a digit
    # This does NOT enforce the label size limits (63 characters max in-between dots)
    regex = re.compile('^[A-Za-z0-9][A-Za-z0-9\-\.]+$')
    if not regex.match(domain):
        return False

    # A domain name MUST end with an alphanumeric character
    # At this point we're certain that the string only contains alphanumeric characters or hyphens and dots.
    # So we just need to check that it doesn't end with a hyphen or a dot
    if domain.endswith('-') or domain.endswith('.'):
        return False

    # Domain names have a length limit of 255 characters
    if len(domain) > 255:
        return False

    return True

#If no backup of /etc/hosts was made, make one
if not os.path.exists("/etc/hosts.mintNanny.backup"):
    os.system("cp /etc/hosts /etc/hosts.mintNanny.backup")

#Set the Glade file
gladefile = "/usr/lib/linuxmint/mintNanny/mintnanny.ui"
builder=Gtk.Builder()
builder.add_from_file(gladefile)
window = builder.get_object("window1")
window.set_title(_("Domain Blocker"))
vbox = builder.get_object("vbox_main")
treeview_domains = builder.get_object("treeview_domains")
window.set_icon_from_file("/usr/lib/linuxmint/mintNanny/icon.svg")

# the treeview
column1 = Gtk.TreeViewColumn(_("Blocked domains"), Gtk.CellRendererText(), text=0)
column1.set_sort_column_id(0)
column1.set_resizable(True)
treeview_domains.append_column(column1)
treeview_domains.set_headers_clickable(True)
treeview_domains.set_reorderable(False)
treeview_domains.show()

model = Gtk.TreeStore(str)
model.set_sort_column_id( 0, Gtk.SortType.ASCENDING )
treeview_domains.set_model(model)

#Get the list of allowed domains
hostsFile = open("/etc/hosts")
for line in hostsFile:
    line = str.strip(line)
    if line.find('0.0.0.0') > -1 or line.find('blocked by mintNanny') > -1:
        elements = line.split("\t")
        domain = elements[1]
        iter = model.insert_before(None, None)
        model.set_value(iter, 0, domain)
del model

window.connect("delete_event", Gtk.main_quit)
builder.get_object("toolbutton_add").connect("clicked", add_domain, treeview_domains, window)
builder.get_object("toolbutton_remove").connect("clicked", remove_domain, treeview_domains)

window.show_all()

Gtk.main()



usr/lib/linuxmint/mintnanny/mintnanny.py
#!/usr/bin/python3

# MintNanny
#   Clement Lefebvre <clem@linuxmint.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; Version 2
# of the License.

import os
import re
import sys
import gettext
import fileinput
import gi
gi.require_version("Gtk", "3.0")
from gi.repository import Gtk, Gdk

# i18n
gettext.install("mintnanny", "/usr/share/linuxmint/locale")


class MintNanny():

    def __init__(self):
        # Set the Glade file
        gladefile = "/usr/share/linuxmint/mintnanny/mintnanny.ui"
        builder = Gtk.Builder()
        builder.add_from_file(gladefile)
        self.window = builder.get_object("main_window")
        self.window.set_title(_("Domain Blocker"))
        self.window.set_icon_name("mintnanny")

        # the treeview
        column = Gtk.TreeViewColumn(_("Blocked domains"), Gtk.CellRendererText(), text=0)
        column.set_sort_column_id(0)
        column.set_resizable(True)
        self.treeview = builder.get_object("treeview_domains")
        self.treeview.append_column(column)
        self.treeview.set_headers_clickable(True)
        self.treeview.set_reorderable(False)
        self.treeview.show()
        self.model = Gtk.TreeStore(str)
        self.model.set_sort_column_id(0, Gtk.SortType.ASCENDING)
        self.treeview.set_model(self.model)

        # Get the list of allowed domains
        with open("/etc/hosts") as f:
            for line in f:
                if '0.0.0.0' in line or 'blocked by mintNanny' in line:
                    elements = line.split()
                    if len(elements) > 1:
                        domain = elements[1]
                        iter = self.model.insert_before(None, None)
                        self.model.set_value(iter, 0, domain)

        self.window.connect("delete_event", Gtk.main_quit)
        builder.get_object("toolbutton_add").connect("clicked", self.add_domain)
        self.remove_button = builder.get_object("toolbutton_remove")
        self.remove_button.connect("clicked", self.remove_domain)
        self.remove_button.set_sensitive(False)
        self.treeview.get_selection().connect("changed", self.on_domain_selected)

        self.window.show_all()

    def ask_domain_name(self):
        dialogWindow = Gtk.MessageDialog(self.window, Gtk.DialogFlags.MODAL | Gtk.DialogFlags.DESTROY_WITH_PARENT, Gtk.MessageType.QUESTION, Gtk.ButtonsType.OK_CANCEL, _("Please type the domain name you want to block"))
        dialogWindow.set_title(_("Domain name"))
        dialogBox = dialogWindow.get_content_area()
        entry = Gtk.Entry()
        entry.set_activates_default(True)
        box = Gtk.Box()
        box.pack_start(entry, True, True, 12)
        dialogBox.pack_start(box, True, True, 0)
        okButton = dialogWindow.get_widget_for_response(response_id=Gtk.ResponseType.OK)
        okButton.set_can_default(True)
        okButton.grab_default()
        dialogWindow.show_all()
        response = dialogWindow.run()
        text = entry.get_text()
        dialogWindow.destroy()
        if (response == Gtk.ResponseType.OK) and (text != ''):
            return text
        else:
            return None

    def add_domain(self, widget):
        domain = self.ask_domain_name()
        if domain is None or domain == '':
            # Take no action on empty input
            return
        domain = re.sub(r'\s', '', domain)
        if domain.startswith('www.'):
            domain = domain[4:]
        if not self.is_valid_domain(domain):
            # User has passed an invalid domain (one that contains invalid characters)
            # Display an error dialog to inform them why we're not adding it to the list
            dlg = Gtk.MessageDialog(type=Gtk.MessageType.ERROR, buttons=Gtk.ButtonsType.OK, message_format=_("Invalid Domain"))
            dlg.set_transient_for(self.window)
            desc1 = _("%s is not a valid domain name." % domain)
            desc2 = _("Domain names must start and end with a letter or a digit, and can only contain letters, digits, dots and hyphens.")
            desc3 = _("Example: my.number1domain.com")
            dlg.format_secondary_text("%s\n\n%s\n\n%s" % (desc1, desc2, desc3))
            dlg.run()
            dlg.destroy()
            return

        prefixes = [""]
        if len(domain.split(".")) == 2:
            # domain in the form 'domainname.extension'
            prefixes.append("www.")
        for prefix in prefixes:
            full_domain = "%s%s" % (prefix, domain)
            iter = self.model.insert_before(None, None)
            self.model.set_value(iter, 0, full_domain)
            os.system("echo \"127.0.0.1 %s # blocked by mintNanny\" >> /etc/hosts" % full_domain)

    def on_domain_selected(self, selection):
        model, treeiter = selection.get_selected()
        self.remove_button.set_sensitive(treeiter != None)

    def remove_domain(self, widget):
        selection = self.treeview.get_selection()
        (model, iter) = selection.get_selected()
        if (iter != None):
            domain = model.get_value(iter, 0)
            for line_number, line in enumerate(fileinput.input('/etc/hosts', inplace=1)):
                found = False
                if '0.0.0.0' in line or 'blocked by mintNanny' in line:
                    elements = line.split()
                    if len(elements) > 1:
                        if elements[1] == domain:
                            found = True
                if not found:
                    sys.stdout.write(line)
            model.remove(iter)

    def is_valid_domain(self, domain):
        # Quick sanity check
        if domain == '' or "." not in domain:
            return False

        # The following is based on RFC 952 (https://tools.ietf.org/html/rfc952)
        # and section 2.1 of RFC 1123 (https://tools.ietf.org/html/rfc1123#page-13)
        # Also see sections 2.3.1 and 2.3.4 of RFC 1035 (http://tools.ietf.org/html/rfc1035)

        # Quick regex match to check the domain name's sanity
        # Note: This enforces that the domain name starts with a letter or a digit
        # This does NOT enforce the label size limits (63 characters max in-between dots)
        regex = re.compile('^[A-Za-z0-9][A-Za-z0-9\-\.]+$')
        if not regex.match(domain):
            return False

        #Enforces label size of 63 characters or less
        tokens = domain.split('.')
        for token in tokens:
            if len(token) > 63:
                return False

        # A domain name MUST end with an alphanumeric character
        # At this point we're certain that the string only contains alphanumeric characters or hyphens and dots.
        # So we just need to check that it doesn't end with a hyphen or a dot
        if domain.endswith('-') or domain.endswith('.'):
            return False

        # Domain names have a length limit of 255 characters
        if len(domain) > 255:
            return False

        return True

if __name__ == "__main__":
    # If no backup of /etc/hosts was made, make one
    if not os.path.exists("/etc/hosts.mintnanny.backup"):
        os.system("cp /etc/hosts /etc/hosts.mintnanny.backup")
    MintNanny()
    Gtk.main()
